File: import_employees.py
#!/usr/bin/env python3
import csv
import requests
import json
from datetime import datetime

# Admin credentials
EMAIL = "mariamy@inara.org"
PASSWORD = "inara2024"
BASE_URL = "http://localhost:8000/api/v1"

def login():
    """Login and get access token"""
    print("Logging in...")
    response = requests.post(
        f"{BASE_URL}/auth/login",
        data={
            "username": EMAIL,
            "password": PASSWORD
        }
    )
    
    if response.status_code == 200:
        data = response.json()
        token = data.get("data", {}).get("access_token")
        if token:
            print("✓ Login successful\n")
            return token
    
    print(f"✗ Login failed: {response.text}")
    return None

def convert_date(date_str):
    """Convert DD/MM/YYYY to YYYY-MM-DD"""
    if not date_str or not date_str.strip():
        return None
    
    try:
        # Handle DD/MM/YYYY format
        if '/' in date_str:
            parts = date_str.split('/')
            if len(parts) == 3:
                day = parts[0].zfill(2)
                month = parts[1].zfill(2)
                year = parts[2] if len(parts[2]) == 4 else f"20{parts[2]}"
                return f"{year}-{month}-{day}"
    except:
        pass
    
    return None

def import_employee(employee, token):
    """Import a single employee"""
    headers = {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json"
    }
    
    response = requests.post(
        f"{BASE_URL}/employees/",
        headers=headers,
        json=employee
    )
    
    return response

def main():
    # Login first
    token = login()
    if not token:
        return
    
    # Read CSV file
    print("Reading CSV file...")
    employees = []
    
    with open('/Users/maiwand/INARA-HR/final.csv', 'r', encoding='utf-8') as file:
        csv_reader = csv.DictReader(file)
        
        for row in csv_reader:
            # Skip rows without required fields
            if not row.get('First Name') or not row.get('Last Name'):
                continue
            
            # Map CSV fields to API fields
            employee = {}
            
            if row.get('Employee Number'):
                employee['employee_number'] = row['Employee Number'].strip()
            
            if row.get('First Name'):
                employee['first_name'] = row['First Name'].strip()
            
            if row.get('Last Name'):
                employee['last_name'] = row['Last Name'].strip()
            
            if row.get('Work Email'):
                employee['work_email'] = row['Work Email'].strip()
            
            if row.get('Phone'):
                employee['phone'] = row['Phone'].strip()
            
            if row.get('Mobile'):
                employee['mobile'] = row['Mobile'].strip()
            
            if row.get('Date of Birth (YYYY-MM-DD)'):
                date = convert_date(row['Date of Birth (YYYY-MM-DD)'])
                if date:
                    employee['date_of_birth'] = date
            
            if row.get('Gender (male/female/other)'):
                gender = row['Gender (male/female/other)'].strip().lower()
                if gender in ['male', 'female', 'other']:
                    employee['gender'] = gender
            
            if row.get('Nationality (Country Code)'):
                employee['nationality'] = row['Nationality (Country Code)'].strip()
            
            if row.get('Employment Type (full_time/part_time/consultant/volunteer)'):
                emp_type = row['Employment Type (full_time/part_time/consultant/volunteer)'].strip().lower().replace(' ', '_')
                if emp_type in ['full_time', 'part_time', 'consultant', 'volunteer']:
                    employee['employment_type'] = emp_type
            
            if row.get('Work Location'):
                employee['work_location'] = row['Work Location'].strip()
            
            if row.get('Hire Date (YYYY-MM-DD)'):
                date = convert_date(row['Hire Date (YYYY-MM-DD)'])
                if date:
                    employee['hire_date'] = date
            
            if row.get('Position/Title'):
                employee['position'] = row['Position/Title'].strip()
            
            if row.get('Department'):
                employee['department'] = row['Department'].strip()
            
            # Only add if we have required fields
            if employee.get('first_name') and employee.get('last_name'):
                employees.append(employee)
    
    print(f"Found {len(employees)} employees to import\n")
    
    # Import employees
    success_count = 0
    failed_count = 0
    errors = []
    
    for i, employee in enumerate(employees, 1):
        name = f"{employee.get('first_name', '')} {employee.get('last_name', '')}"
        print(f"[{i}/{len(employees)}] Importing {name}...", end=' ')
        
        response = import_employee(employee, token)
        
        if response.status_code in [200, 201]:
            success_count += 1
            print("✓")
        else:
            failed_count += 1
            print("✗")
            try:
                error_data = response.json()
                error_msg = error_data.get('error', {}).get('message') or error_data.get('detail') or response.text
            except:
                error_msg = response.text[:200]
            
            errors.append({
                'name': name,
                'email': employee.get('work_email', 'N/A'),
                'error': error_msg
            })
    
    # Summary
    print("\n" + "=" * 50)
    print("Import Summary:")
    print(f"✓ Successful: {success_count}")
    print(f"✗ Failed: {failed_count}")
    print("=" * 50)
    
    if errors:
        print("\nErrors:")
        for idx, err in enumerate(errors[:20], 1):  # Show first 20 errors
            print(f"\n{idx}. {err['name']} ({err['email']})")
            print(f"   Error: {err['error'][:200]}")
        
        if len(errors) > 20:
            print(f"\n... and {len(errors) - 20} more errors")

if __name__ == "__main__":
    main()

File: import_employees.js
const fs = require('fs');
const https = require('https');
const http = require('http');

// Read the CSV file
const csvData = fs.readFileSync('/Users/maiwand/INARA-HR/final.csv', 'utf-8');

// Parse CSV
function parseCSV(text) {
  const lines = text.split('\n').filter(line => line.trim());
  if (lines.length < 2) return [];

  const headers = lines[0].split(',');
  const employees = [];

  for (let i = 1; i < lines.length; i++) {
    const values = parseCSVLine(lines[i]);
    if (values.length === 0) continue;

    const employee = {};
    headers.forEach((header, index) => {
      const value = values[index]?.trim() || '';
      
      switch (header.trim()) {
        case 'First Name':
          employee.first_name = value;
          break;
        case 'Last Name':
          employee.last_name = value;
          break;
        case 'Work Email':
          employee.work_email = value;
          break;
        case 'Phone':
          employee.phone = value;
          break;
        case 'Mobile':
          employee.mobile = value;
          break;
        case 'Date of Birth (YYYY-MM-DD)':
          employee.date_of_birth = convertDate(value);
          break;
        case 'Gender (male/female/other)':
          employee.gender = value.toLowerCase();
          break;
        case 'Nationality (Country Code)':
          employee.nationality = value;
          break;
        case 'Employment Type (full_time/part_time/consultant/volunteer)':
          employee.employment_type = value.toLowerCase().replace(' ', '_');
          break;
        case 'Work Location':
          employee.work_location = value;
          break;
        case 'Hire Date (YYYY-MM-DD)':
          employee.hire_date = convertDate(value);
          break;
        case 'Position/Title':
          employee.position = value;
          break;
        case 'Department':
          employee.department = value;
          break;
        case 'Employee Number':
          employee.employee_number = value;
          break;
      }
    });

    // Only add if we have required fields
    if (employee.first_name && employee.last_name && employee.work_email) {
      employees.push(employee);
    }
  }

  return employees;
}

function parseCSVLine(line) {
  const values = [];
  let current = '';
  let inQuotes = false;

  for (let i = 0; i < line.length; i++) {
    const char = line[i];
    
    if (char === '"') {
      inQuotes = !inQuotes;
    } else if (char === ',' && !inQuotes) {
      values.push(current);
      current = '';
    } else {
      current += char;
    }
  }
  
  values.push(current);
  return values;
}

function convertDate(dateStr) {
  if (!dateStr || dateStr.trim() === '') return null;
  
  // Handle DD/MM/YYYY format
  if (dateStr.includes('/')) {
    const parts = dateStr.split('/');
    if (parts.length === 3) {
      const day = parts[0].padStart(2, '0');
      const month = parts[1].padStart(2, '0');
      const year = parts[2].length === 2 ? '20' + parts[2] : parts[2];
      return `${year}-${month}-${day}`;
    }
  }
  
  return dateStr;
}

async function importEmployee(employee, token) {
  return new Promise((resolve, reject) => {
    const data = JSON.stringify(employee);
    
    const options = {
      hostname: 'localhost',
      port: 8000,
      path: '/api/v1/employees/',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
        'Content-Length': data.length
      }
    };

    const req = http.request(options, (res) => {
      let responseData = '';

      res.on('data', (chunk) => {
        responseData += chunk;
      });

      res.on('end', () => {
        if (res.statusCode === 200 || res.statusCode === 201) {
          resolve({ success: true, data: responseData });
        } else {
          resolve({ success: false, error: responseData, status: res.statusCode });
        }
      });
    });

    req.on('error', (error) => {
      resolve({ success: false, error: error.message });
    });

    req.write(data);
    req.end();
  });
}

async function main() {
  // You need to replace this with your actual access token
  const token = process.argv[2];
  
  if (!token) {
    console.error('Usage: node import_employees.js <access_token>');
    console.error('\nTo get your access token:');
    console.error('1. Log in to the INARA HR system');
    console.error('2. Open browser console (F12)');
    console.error('3. Type: localStorage.getItem("access_token")');
    console.error('4. Copy the token and run: node import_employees.js <token>');
    process.exit(1);
  }

  console.log('Parsing CSV file...');
  const employees = parseCSV(csvData);
  console.log(`Found ${employees.length} employees to import\n`);

  let successCount = 0;
  let failedCount = 0;
  const errors = [];

  for (let i = 0; i < employees.length; i++) {
    const employee = employees[i];
    process.stdout.write(`[${i + 1}/${employees.length}] Importing ${employee.first_name} ${employee.last_name}...`);
    
    const result = await importEmployee(employee, token);
    
    if (result.success) {
      successCount++;
      console.log(' ✓');
    } else {
      failedCount++;
      console.log(' ✗');
      errors.push({
        name: `${employee.first_name} ${employee.last_name}`,
        email: employee.work_email,
        error: result.error,
        status: result.status
      });
    }
    
    // Small delay to avoid overwhelming the server
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  console.log('\n' + '='.repeat(50));
  console.log('Import Summary:');
  console.log(`✓ Successful: ${successCount}`);
  console.log(`✗ Failed: ${failedCount}`);
  console.log('='.repeat(50));

  if (errors.length > 0) {
    console.log('\nErrors:');
    errors.forEach((err, idx) => {
      console.log(`\n${idx + 1}. ${err.name} (${err.email})`);
      console.log(`   Status: ${err.status}`);
      console.log(`   Error: ${err.error.substring(0, 200)}`);
    });
  }
}

main().catch(console.error);
